File: Concurrency/1114.按序打印/Foo.py
from threading import Lock

class Foo:
    def __init__(self):
        self.first_job_done = Lock()
        self.second_job_done = Lock()
        self.first_job_done.acquire()
        self.second_job_done.acquire()

    def first(self, printFirst: 'Callable[[], None]') -> None:
        # printFirst() outputs "first". Do not change or remove this line.
        printFirst()
        self.first_job_done.release()

    def second(self, printSecond: 'Callable[[], None]') -> None:
        with self.first_job_done:
            # printSecond() outputs "second". Do not change or remove this line.
            printSecond()
            self.second_job_done.release()

    def third(self, printThird: 'Callable[[], None]') -> None:
        with self.second_job_done:
            # printThird() outputs "third". Do not change or remove this line.
            printThird()

File: Concurrency/1114/Foo.java
class Foo {
    private Semaphore firstJobDone = new Semaphore(0);
    private Semaphore secondJobDone = new Semaphore(0);

    public void first(Runnable printFirst) throws InterruptedException {
        // printFirst.run() outputs "first". Do not change or remove this line.
        printFirst.run();
        firstJobDone.release();
    }

    public void second(Runnable printSecond) throws InterruptedException {
        firstJobDone.acquire();
        // printSecond.run() outputs "second". Do not change or remove this line.
        printSecond.run();
        secondJobDone.release();
    }

    public void third(Runnable printThird) throws InterruptedException {
        secondJobDone.acquire();
        // printThird.run() outputs "third". Do not change or remove this line.
        printThird.run();
    }
}
